### PR TITLE
optimizations and other improvements 

### PR DESCRIPTION
**Pull request by dbenson24**
#71 

Opening the PR, it's definitely not ready to be merged yet there's a couple more things I'm finishing up but I wanted to open earlier so if you wanted to read things over you'd be able to

### CODE REVIEW

File: bvh-lib/Cargo.toml
Comment (svenstaro) [inline bvh-lib/Cargo.toml:25]: Do we want debug here?

File: bvh-lib/Cargo.toml
Comment (svenstaro) [inline bvh-lib/Cargo.toml:30]: Do we want debug here?

**Review by svenstaro — CHANGES_REQUESTED** [review]

**Comment by svenstaro** [comment]
Is there any chance you could cut down the PR into smaller chunks? I realize you did a lot of refactoring but are there perhaps any smaller chunks to make reviewing easier?

**Comment by dbenson24** [comment]
The FFI stuff can very easily be separated so I'll do that. I'm not even sure what the best way to distribute that is

I think I can also pull the new shapes out into a separate change, I'll see what I can come up with

File: bvh-f64/Cargo.toml
Comment (svenstaro) [inline bvh-f64/Cargo.toml:2]: Do we really need a specialized crate for this? Can't we use traits for this to have Rust do the specialization?

**Review by svenstaro — CHANGES_REQUESTED** [review]

File: flamegraph.svg
Comment (svenstaro) [inline flamegraph.svg:1]: Did you mean to commit this file? A Makefile target might be more appropriate to generate this or perhaps even just a README entry.

**Review by svenstaro — CHANGES_REQUESTED** [review]

File: flamegraph.svg
Comment (dbenson24) [reply flamegraph.svg:1]: Yea I thought I had already deleted this

File: bvh-f64/Cargo.toml
Comment (dbenson24) [reply bvh-f64/Cargo.toml:2]: So I tried a few approaches to this. It might be a possibility with nalgebra backing things to have the float be a generic on everything, but with glam there's no trait to abstract dvec3 from vec3 so we would have to wrap that ourselves. This trick using separate crates in the workspace is something I got from the author of nalgebra and how he builds rapier (2d 3d 32 and 64bit). When the workspace is all setup with this cargo test runs the test for all the versions and you'll get compiler errors from rust analyzer for both at the same time

**Comment by dbenson24** [comment]
I've been a bit busy last couple of weeks haven't had time to look at breaking this apart. if that's still something you would prefer I can start working on trying to do so

**Comment by svenstaro** [comment]
Well if we can somehow do this in small incremental PRs that would be ideal but I can understand if you don't want to go down that path but it would make reviewing faster for me for sure. You get to decide. :)

**Comment by dbenson24** [comment]
I've pulled the ffi bindings out, I can add them back in if you want them in the repo otherwise I will just throw them up in a separate repo.

In terms of reviewing this, the big change that affects the whole crate is refactoring all the types to use the vec/float types defined at the crate root. Otherwise the changes are pretty much file by file but because of that change to all of the types it's kind of a pain to break this down into little commits.

File: src/shapes/capsule.rs
Comment (Elabajaba) [inline src/shapes/capsule.rs:41]: Commented out code. Delete it if it isn't used?

File: src/shapes/ray.rs
Comment (Elabajaba) [inline src/shapes/ray.rs:384]: More commented out code.

File: src/shapes/ray.rs
Comment (Elabajaba) [inline src/shapes/ray.rs:81]: Do you have your own IntersectionRay impls outside of this library? I'm wondering what the performance tradeoff of using a vtable here vs an enum would be.

Also, is there a reason why IntersectionRay is only implemented for sphere+triangle, but not aabb/capsule/obb?

File: bvh-f64/Cargo.toml
Comment (Elabajaba) [inline bvh-f64/Cargo.toml:48]: This was stabilized in Rust 1.60.

File: bvh/Cargo.toml
Comment (Elabajaba) [inline bvh/Cargo.toml:46]: This was stabilized in Rust 1.60.

File: src/mesh/convex_hull.rs
Comment (Elabajaba) [inline src/mesh/convex_hull.rs:1]: Empty file.

File: src/mesh/mod.rs
Comment (Elabajaba) [inline src/mesh/mod.rs:1]: Also empty.

**Comment by Elabajaba** [comment]
So I decided to try building a toy pathtracer against this branch, and I'm curious about how users are supposed to use the included shapes? 

Currently none of them can actually be used in the BVH since none of them have a BHShape impl (previously you only needed a Bounded impl, in this branch you need both a Bounded and a BHShape impl), so you can't actually include them in the BVH. If you're supposed to shoot rays and intersect them against the included shapes, it'd be nice if they used a Structure of Arrays (SoA) layout (a simple SoA layout gets me ~2x the performance with 100 spheres, or ~3x with 1000 spheres in my pathtracer), however because intersects_ray (from the IntersectionRay impl) doesn't have a way to return an index, you can't currently use your own SOA layouts.

I believe IntersectionAABB also doesn't have any way to return an index, so it also can't be used with SOA layouts.

File: src/testbase.rs
Comment (Elabajaba) [inline src/testbase.rs:168]: I feel like this should be removed since this pr includes a triangle shape, and having 2 public triangle structs is confusing.

**Comment by dbenson24** [comment]
@Elabajaba the shapes missing impls are just because I haven't had time to write that code. This PR stems entirely from me writing other code and adding stuff to BVH as I needed it. In terms of the shapes not being able to be used in SOA, I chose not to attach an index to any of the shapes because it's very easy if you want to do that yourself. Just define an IndexedShape<T> that stores the shape and it's index. I didn't think that it made sense to force an index onto every shape.

File: bvh-f64/Cargo.toml
Comment (marstaik) [reply bvh-f64/Cargo.toml:2]: I would very much like to see this move to nalgebra, it would make the usage much cleaner via `Bvh<T>`. This would definitely be possible with the nalgebra vector types and number bounds.

Heck, because of the AABB's, assuming there are no dot products/cross products (which I don't think there are - haven't checked), it should work with integer types as well, and maybe even unsigned types (but there would have to be better bounds checking).

I could try and help get this working with nalgebra as well.

Comment (svenstaro) [reply bvh-f64/Cargo.toml:2]: I think that would be pretty cool.

Comment (marstaik) [reply bvh-f64/Cargo.toml:2]: I'm not sure how to approach this for now though, should I wait for these changes to get merged in?

Comment (marstaik) [reply bvh-f64/Cargo.toml:2]: Hmmm it might be easier for me to port it now, and have this PR rebase on it. Otherwise we would be splitting code and then remerging.

Comment (Elabajaba) [reply bvh-f64/Cargo.toml:2]: Has nalgebra fixed the issues that made this crate switch from nalgebra to glam previously?

**Comment by svenstaro** [comment]
Because of the discussion in the other thread: I wouldn't mind a co-maintainer on this. I don't use it much myself currently and I think someone with more stakes in this could really drive it forward a bit.

**Comment by marstaik** [comment]
> Because of the discussion in the other thread: I wouldn't mind a co-maintainer on this. I don't use it much myself currently and I think someone with more stakes in this could really drive it forward a bit.

I may be interested, I plan to use it quite a bit in the future. I am a bit new to the rust side of things, but I generally worked with c++ metaprogramming before my switch to game engine development in rust.

**Comment by svenstaro** [comment]
> > Because of the discussion in the other thread: I wouldn't mind a co-maintainer on this. I don't use it much myself currently and I think someone with more stakes in this could really drive it forward a bit.
> 
> I may be interested, I plan to use it quite a bit in the future. I am a bit new to the rust side of things, but I generally worked with c++ metaprogramming before my switch to game engine development in rust.

How about you start off by making one or two PRs that I'll review and then I can make you contrib from there if we click well?

**Comment by svenstaro** [comment]
@marstaik how do you wanna play this PR? Shall we close it? Shall we leave it open until you have implemented the bits that you are interested in? I don't think the original author is going to return to this any time soon and the rebase is very gnarly at this point.

**Comment by marstaik** [comment]
I'm going to close this, the original author has already started porting the changes incrementally starting with this:
https://github.com/svenstaro/bvh/pull/99

**Pull request closed without merging**
